src/Workspaces/Core/Portable/Utilities/ValuesSources/OptionalValueSource.cs

```diff
@@ -0,0 +1,22 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+#nullable enable
+
+using System.Diagnostics.CodeAnalysis;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Roslyn.Utilities
+{
+    /// <summary>
+    /// A class that abstracts the accessing of a value that might not be available anymore when queried.
+    /// </summary>
+    internal abstract class OptionalValueSource<T> where T : class
```

Comment:
💭 It seems like we could make `ValueSource<T>` extend this class with the following signature changes here:

```
public abstract bool TryGetValue([MaybeNullWhen(false)] out T value);
public abstract Optional<T> GetOptionalValue(CancellationToken cancellationToken);
public abstract Task<Optional<T>> GetOptionalValueAsync(CancellationToken cancellationToken);
```

Then `ValueSource<T>` would implement the above and provide non-optional equivalents:

```
public abstract T GetValue(CancellationToken cancellationToken);
public abstract Task<T> GetValueAsync(CancellationToken cancellationToken);
```